Modularize dependency on vdrtools/indy with option for indy-Vdr and Indy-Credx in place

gmulhearn-anonyome opened the pull request:
# Summary of Changes
* Created `Profile`, `BaseWallet`, `BaseLedger`, and `BaseAnoncreds` traits/interfaces. Where Wallet Ledger and Anoncreds APIs signatures are based mostly on indy/vdrtools APIs for the respective sections. A Profile represents a collection of these 3 traits/dependencies bundled together - Profile is derived from ACA-py's approach.
* Created "IndySdk" (ACA-py terms) implementations of Wallet/Ledger/Anoncreds - which mostly just calls the `/indy` directory functionality directly.
* Created an `IndySdkProfile` implementation which bundles together those 3 Indy impls ^.
* Created `IndyVdr` implementation of Ledger, which uses indy-vdr for ledger request/response formation and sending. The profile's wallet is used for signing transactions. The implementation of most functions is derived from ACA-py.
* Created `IndyCredx` implementation of Anoncreds, which uses indy-credx for the formation of anoncreds related objects - e.g. creating proofs and link secrets. The profile's wallet is used for storing/searching credentials and handling master secrets. The implementation of most functions is derived from ACA-py, especially around the storage of creds as wallet records.
* Created a `ModularWalletProfile` which bundles vdr Ledger and credx Anoncreds with any user provided `BaseWallet` (I.e. a pluggable wallet). This is similar to ACA-py's `AskarProfile` - which we may wish to impl in the future.
* Updated integration tests (where possible) to dynamically pick between using an `IndySdk` or `ModularWallet` Profile (by setting the `modular_deps` flag)
* Created a `MockProfile` which is used by unit tests. Currently MockProfile contains Mock[Ledger|Wallet|Anoncreds], all of which return the mock data which you'd expect to be returned if indy_mocks/did_mocks were enabled.
* Add a new CI step to rerun integration tests with coverage using the 'modular_profile' flag, which will trigger ~50 int tests to run using the modularWalletProfile instead of IndySdkProfile. (note that in the future ALL tests should be integration tests should be triggered by this flag - currently only 50 due to missing implementations)

# Future Work
* IndyVdrLedger and IndyCredx implementations were made with a focus on Holder/Prover functionality. As such, some 'institutional/verifier/issuer' functionality has stubbed implementations and needs to be populated.
* Libvcx would ideally support the ModularProfile, not just default to Indy. More thought and planning will be needed
* May wish to move `indy` functions into the `IndySdkWallet/Ledger/Anoncreds` files directly, as the `indy` directory is somewhat thin now.
* AskarWallet and AskarProfile implementation
* May wish to consider creating interface functions for Wallet Management APIs (i.e. generic create/open/close/delete/export/etc APIs)
* Consider a better mechanism for integration testing both versions of Profile (i.e. Modular and Indy), rather than relying on the CI running the job twice.
* Unit testing on credx/indyvdr?
* `MockProfile` may open up a good opportunity to make some mocking much easier. E.g. the injection logic could be built into the MockProfile methods rather than relying on global settings. e.g. `mock_profile.set_verifier_verify_result(true)`, etc
* switch all integration tests to optionally use the ModularWalletProfile
* switch indy-credx to anoncreds-rs once it is released/stable https://github.com/hyperledger/anoncreds-rs

# Concerns
* The Credx implementation stores Link Secrets in the wallet as records with a "LINK SECRET" type. An external consumer could access these records, therefore leaking the Link Secret.


# Architecture (Before (TOP) and After (BOTTOM))
![VCXIndyDepsArchitecture drawio](https://user-images.githubusercontent.com/90162009/205196716-8fff808e-3e25-4f4a-9ea3-0af8c46ecef5.png)

gmulhearn-anonyome commented on aries_vcx/src/error.rs:
```diff
@@ -568,6 +584,9 @@ impl From<VcxErrorKind> for u32 {
             VcxErrorKind::CreatePublicAgent => error::CREATE_PUBLIC_AGENT.code_num,
             VcxErrorKind::CreateOutOfBand => error::CREATE_OUT_OF_BAND.code_num,
             VcxErrorKind::CreateAgent => error::CREATE_AGENT.code_num,
+            VcxErrorKind::InvalidInput => error::INDY_VDR_INPUT_INPUT.code_num,
```
todo - fix error:: name

gmulhearn-anonyome reviewed (commented):
issue: test-android-build failure in cargo test is not triggering a failure of the pipeline

gmulhearn-anonyome commented on .github/workflows/main.yml:
```diff
@@ -242,6 +242,36 @@ jobs:
+
 #  #  ##########################################################################################
 #  #  ###############################    TESTING    ###########################################
 
```
issue; test-android-build is not failing despite cargo test failing

Patrik-Stas replied:
I suppose this is solved now?

Patrik-Stas commented on .vscode/settings.json:
```diff
@@ -0,0 +1,15 @@
+{
```
seems like this was commited before `vscode` was put in `.gitignore`

gmulhearn-anonyome replied:
thnx, have removed this now

Patrik-Stas commented on aries_vcx/Cargo.toml:
```diff
@@ -31,11 +33,16 @@ serde = "1.0.97"
 libvdrtools = { git = "https://gitlab.com/left-arm/vdr-tools.git", rev = "debf776c8" }
+# indy alternatives ----
+indy-vdr = "0.3.4"
+indy-credx = { git = "https://github.com/anonyome/indy-shared-rs.git", rev = "7342bc624d23ece8845d1a701cd2cdc9cd401bb0" }
```
What kind of changes has been done in your fork? Do you plan to open PR?

gmulhearn-anonyome replied:
i've added a lengthy comment in the toml justifying this fork, as well as instructions on how to monitor it / resolve in the future. it's unfortunate that we have to use a fork, but let me know if you can think of a better temporary solution

Patrik-Stas commented on aries_vcx/Cargo.toml:
```diff
@@ -31,11 +33,16 @@ serde = "1.0.97"
 serde_json = "1.0.40"
 serde_derive = "1.0.97"
 regex = "1.1.0"
+reqwest = "0.11.10"
```
Can't see usage of this

(btw unrelated to your PR - just FYI, we are somewhat misusing `agency-client/src/httpclient.rs`, would like to get rid of that)

gmulhearn-anonyome replied:
thnx, deleted this

gmulhearn-anonyome commented on .github/workflows/main.yml:
```diff
@@ -242,6 +242,36 @@ jobs:
           name: code-coverage-report-unit-aries-vcx
           path: /tmp/artifacts/coverage
 
+  #TODO - can this be included within code-coverage-aries-vcx-integration-tests?
```
@Patrik-Stas any thoughts on this? not too familiar with best practices for your github pipelines

Patrik-Stas replied:
as discussed on today's aries-vcx call, good for now, we'll refactor coverege handling across the board in CI as separate task

gmulhearn-anonyome commented on aries_vcx/src/core/profile/modular_wallet_profile.rs:
```diff
@@ -0,0 +1,51 @@
+
+#[allow(dead_code)]
+#[derive(Debug)]
+pub struct ModularWalletProfile {
```
Any other naming suggestions for this? The intention is for this to be a profile implementation which is constructed using our new modular deps (indy-vdr + indy-credx) + any `dyn BaseWallet` which you plug in - (i.e. a modular wallet).

Patrik-Stas replied:
`ModularWalletProfile` is okay

gmulhearn-anonyome commented on aries_vcx/src/indy/proofs/prover/mod.rs:
```diff
@@ -29,13 +29,14 @@ pub async fn libindy_prover_create_revocation_state(
             parse_and_validate(rev_reg_def_json)?,
             parse_and_validate(rev_reg_delta_json)?,
-            100,
+            timestamp,
```
is this ok? was wondering why timestamp is hardcoded to 100, I couldn't see any reason for it

Patrik-Stas replied:
this is good, thank you!

gmulhearn-anonyome commented on aries_vcx/src/lib.rs:
```diff
@@ -40,6 +40,10 @@ pub mod global;
 
+pub mod core;
+pub mod plugins;
+pub mod xyz;
```
@Patrik-Stas looking for suggestions on what to call this - lol. As discussed, `xyz` contains a lot of methods previously from `indy` but made to be generic for `Profile`s, not indy handles. They are helper functions and impls which aggregate several calls to profiles

Patrik-Stas replied:
lets do just boring `common` / `misc`

gmulhearn-anonyome replied:
renamed to `common`

gmulhearn-anonyome commented on aries_vcx/src/xyz/proofs/prover/prover_internal.rs:
```diff
@@ -178,15 +173,16 @@ pub async fn build_rev_states_json(pool_handle: PoolHandle, credentials_identifi
+                    tails_file,
                     &rev_reg_def_json,
                     &rev_reg_delta_json,
+                    100, // todo - timestamp as 100?
```
should we be using `timestamp` variable from above?

Patrik-Stas replied:
yes, let's do that

for sake of 3rd party validation, afj is doing that as well
https://github.com/hyperledger/aries-framework-javascript/blob/7f37a6265dcc8310b78c14e59f38f1de19fed8c3/packages/core/src/modules/indy/services/IndyRevocationService.ts#L104

left-arm commented on aries_vcx/Cargo.toml:
```diff
@@ -36,6 +38,22 @@ base64 = "0.10"
 num-traits = "0.2.0"
 libvdrtools = { git = "https://gitlab.com/left-arm/vdr-tools.git", rev = "debf776c8" }
+# vdrtools alternatives ----
+indy-vdr = "0.3.4"
```
I suggest remove feature "zmq_vendored" to fix the android build, something like this:

`indy-vdr = { version = "0.3.4", default-features = false, features = ["ffi", "log"] }`

Patrik-Stas commented:
1. looks good to me, I've verified if some of the older PRs have been correctly merged/rebase into this and didn't find an issue
2. please attach the diagram you were presenting on the aries-vcx call, it could help someone scanning through these changes
3. please do, hopefully last, rebase, and let's get this merged 🚀

gmulhearn-anonyome commented:
Hi @guijd3p, cc @Patrik-Stas , i've rebased your most recent changes for supporting legacy vs non-legacy service endpoint writing and resolving.

I've had to move a few things around, you may want to double check that no regressions have been made - your tests still pass however.

Changes include:
* moving the new write_endpoint, write_endpoint_legacy and get_service out of `indy` and into `common/ledger/transactions`.
* Changing these methods to use a `profile` instead of indy handles.
* Changing these methods to use the profile.ledger for adding and getting attributes, rather than depending on `indy` methods.
* Added a 'clear attributes' step to your 2 tests - I was running into issues where running `test_add_get_service_public` before `test_add_get_service` would cause `test_add_get_service` to fail - as it would resolve the non-legacy service rather than legacy service (which the test is intended to target)

gmulhearn-anonyome commented:
I've squashed commits on this branch for sake of the DCO - rebasing was a struggle.

However commit history can be viewed here: https://github.com/anonyome/aries-vcx/tree/modular_indy

Patrik-Stas commented:
I'll include this https://github.com/hyperledger/aries-vcx/pull/683 in one more release `0.49.0` and then this PR will go to `0.50.0` today or over the weekend.
Nothing else to be merged in the meantime.
